### PR TITLE
Minor reorg for easier testing

### PR DESCRIPTION
**Pull request by mrterry**
Minor reorg to make unit testing providers easier.  Will get to #423 in time.

Makes uri translation function more accessible for tests and adds first pass for a github client that mocks out network communication.

### CODE REVIEW

**Comment by mrterry** [comment]
@bollwyvl This is nearly ready to go.  The tests pass on py2.7 and py3.3.  Something in mock on py3.4 is unhappy and I can't reproduce locally.  It looks like github is throttling me now, so tests are failing.

**Comment by mrterry** [comment]
It looks like bumping the version of mock fixed the previous failure.  Github started throttling me on the last test, but all the github client tests (which use mock) passed.

**Comment by bollwyvl** [comment]
Sounds good! I'll give it a look this weekend!

Really need to get our story together on not relying on GH for most of the
tests... i hate submodules from the PR perspective, but I would hate more
to have to embed xxMB of notebooks to not lose our "tribal knowledge".
probably need to do an instrumentation run and generate a list of all the
things we need.

**Comment by mrterry** [comment]
this PR is a decent start.  you can't end-to-end test everything, but you should be able to fully test each layer.  As I see it there are 4-ish layers:

1. github client
1. the url regex
1. the handler
1. the template rendering

This add decent coverage for the github client and a subset of the url regexes.  In a follow up, I'll add test for some of the handlers.  I'll leave template rendering to someone else.

File: dev-requirements.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 coverage
 invoke
+mock>=1.3.0
```
Comment (bollwyvl) [inline dev-requirements.txt]: Should we just pin everything right now?

File: nbviewer/providers/github/handlers.py [hunk 2]
```diff
@@ -294,8 +294,14 @@ def default_handlers(handlers=[]):
     """Tornado handlers"""
 
     return [
+        # ideally these URIs should have been caught by an appropriate
+        # uri_rewrite rather than letting the url provider catch them and then
+        # fixing it here.
+        # There are probably links in the wild that depend on these, so keep
+        # these handlers for backwards compatibility.
```
Comment (bollwyvl) [inline nbviewer/providers/github/handlers.py]: Good summary

File: nbviewer/utils.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import base64
 import cgi
 import re
+from collections import OrderedDict
```
Comment (bollwyvl) [inline nbviewer/utils.py]: A brave coder. I have tried to stay out of this file!

File: dev-requirements.txt (hunk 1 above)
Comment (mrterry) [inline dev-requirements.txt]: This may be cavalier, but I don't see the need for dev requirements.  The mock requirement is to avoid a bug in mock that affects python 3.4.  I should add a comment describing this.

File: nbviewer/utils.py (hunk 3 above)
Comment (mrterry) [inline nbviewer/utils.py]: There be dragons here.  I have another change in the pipeline that will get the global state out of this module.

**Comment by bollwyvl** [comment]
+1: Having some issues getting it (or anything) running with Docker (just did `fedup` and a few things aren't 100%), so I haven't tested it out there...

**Comment by mrterry** [comment]
cool.  let me know if i can help with anything.  i'm eager to get these changes in so that I can get back to work on the GHE provider.  I have a version deployed internally, but there are few bugs related to gists that will greatly benefit from this.

**Comment by bollwyvl** [comment]
Glad to hear your internal hosting is going well. My group still trying to position ourselves into getting to build a Stash integration.

Perhaps @Carreau or @rgbkrk have a look at this?

**Comment by mrterry** [comment]
Gists need some work and there are still a few regex corner cases, but is
going well.

**Comment by Carreau** [comment]
Well, I never say no to plenty of test.
Also I doubt I still have a huge expertise on nbviewer. 

+1 as far as I can tell.

File: nbviewer/utils.py (hunk 3 above)
Comment (rgbkrk) [inline nbviewer/utils.py]: Hahaha, omg.

**Comment by rgbkrk** [comment]
I say :shipit:
